### ISSUE TITLE
Ignore interfaces when checking if the linearization of inheritance graph is possible

### ISSUE THREAD

**Issue by migoldfinger** [ethereum/solidity#13136]
## Abstract

Following an Interface pattern for every contract leads to unsolvable inheritance situation even if it should not.

## Motivation

I tried to implement a number of base contracts with interfaces. The pattern I was following is that a Interface for a contract should have all visible functions and all events for that contract. For that reason each interface that describes a contract that already derives from an other contract also extends the interface of the derived class.
While this was a success with the base contracts it leads to unsolvable inheritance errors when using this contracts as base for other contracts.

## Specification

Interfaces should not be taken into account when checking the inheritance graph.
Instead since they hold no implementation all interfaces of a contract should be merged to a single interface that hold the functions and events of all used interfaces.
That interface is then used to test if all functions are implemented as needed.

## Backwards Compatibility

To prevent code breaking changes. Interfaces that work that way should be marked with a keyword.

For example: virtual interface <name> is <IName>...

## Current Workaround

This would also reflect the workaround I used at the moment to manually copy methods of base interfaces to the derived one without linking the interfaces with the "is" keyword. The drawback of this workaround is that in big codebases changes in the base interfaces are not reflected in the derived Interfaces without manual intervention.

**Comment by cameel** [comment]
### The feature
I think you have a point. The linearization order matters mostly for two things: deciding which virtual function to call and the order of construction/initialization of base contracts. They both do not matter for interfaces since their functions cannot have implementations and then cannot have constructors or variables that would need to be initialized.

I'm not sure if I'm not missing some corner case but it seems to me like we could indeed relax the rules here and not issue this error for interfaces. The change would not be breaking because it would not affect code generation and would just allow compiling new code that used to result in an error.

### Workaround
Still, even with the current implementation you do not need any wild workarounds to deal with this error as long as you control the inheritance hierarchy. Problems with linearization should always be solvable by rearranging the inheritance lists you have and/or avoiding unnecessary inheritance from the same contract at different levels.

You're getting this error because the order in which you place contracts on the inheritance list imposes some constraints that the compiler must satisfy:
- A contract must must not come before any contract it inherits from.
- A contract must must not come before any of its siblings to the left on the inheritance list.

It would be easier to discuss if you gave an example of a situation you cannot resolve. Since you didn't, let's look at the simplest possible case:
```solidity
interface I {}
contract C is I {}
contract D is C, I {}
```
If you have something like this in your code you're telling the compiler that `C` must come before `I` (because it inherits from `I`) but also `I` must come before `C` (because `C` is before `I` on the inheritance list). This is impossible to satisfy.

You can work around this by either rearranging the inheritance list (`D is I, C`) or cleaning it up so that `I` does not appear in the hierarchy twice (`D is C`).

For contracts doing this might affect contract behavior but in case of interfaces it's always safe (which is also the reason why the feature you're requesting would be possible at all).

**Comment by migoldfinger** [comment]
@cameel  thanks for considering
I will try to recreate the problem. I already moved forward so I will have to recreate some from my memory
The Example is loosely based on openzeppelin contracts ERC20. Lets assume they have an interface for each contract even if they do not have it, because that's basically what I have done.
```
interface IERC20 {}
interface IOwnable {}
interface IERC20AltApprove {} // holds decreaseAllowance and increaseAllowance functions.
interface IERC20Metadata is IERC20 {} // optional metadata elements.
interface IERC20Burnable is IERC20Metadata {}
abstract contract Context {}
abstract contract Ownable is IOwnable, Context {}
contract ERC20 is IERC20Metadata, IERC20AltApprove, Context {}
contract ERC20Burnable is IERC20Burnable, ERC20, Context {}
```
I got this setting working without error messages.
The unsolvable problem pops up if I want to use this to create my own token.
```
interface IMyToken is IOwnable, IERC20Burnable, IERC20AltApprove {}
contract MyToken, IMyToken, Ownable, ERC20Burnable {}
```
I think I already tried each and every possible combination of ordering without success.
As far as I understand that this is the case because for example IERC20Burnable interface is used at ERC20Burnable and in MyToken via IMyToken and the compiler can not resolve this circle reference.

**Comment by cameel** [comment]
It must have been slightly different because this
> ```solidity
> contract ERC20Burnable is IERC20Burnable, ERC20, Context {}
> ```

actually looks like this in OZ
> ```solidity
> contract ERC20Burnable is IERC20Burnable, Context, ERC20 {}
> ```

and with that change there's no linearization error.

**Comment by migoldfinger** [comment]
@cameel I know, I can and did solve the linearization error, if any, for the base classes.

The difference in my order is because I wrote the example out of memory and I tried many alternative orders in my attempt so solvle the error at the MyToken class.
The at the moment unsolveable linearization error happens at MyToken.
Even if the base classes compile fine.
There is no way to get MyToken working in my example.

Also, I am not sure if this is a separated feature or related. If you use interfaces in that way you came across cases where
an Interface holds a function that is defined in the interface and also already implemented in some base class.
If that happens you are forced to override the function and call the implementation for Example

IMyToken contains `function allowance(address owner, address spender) external view returns (uint256);`
This function is also present in `IERC20` and implemented in `ERC20` and inherited by `ERC20Burnable`
```
contract MyToken is IMyToken, ERC20Burnable, Ownable
{
...
	function allowance(address owner, address spender) override (IMyToken, ERC20) public view returns (uint256)
	{
		return ERC20.allowance(owner, spender);
	}
...
}
```
I think it is save to assume that if only one implementation is present ether by direct implementation or by inheritance that implementation should be used by default without the need to explicitly call the implementation like in the code above.
Another drawback I came over my workaround is that while you can define functions and override them in the way shown you can not do the same thing with events. You will just get an `DeclarationError: Event with same name and parameter types defined twice.` error.

**Comment by cameel** [comment]
> The difference in my order is because I wrote the example out of memory and I tried many alternative orders in my attempt so solvle the error at the MyToken class.
> The at the moment unsolveable linearization error happens at MyToken.
> Even if the base classes compile fine.
> There is no way to get MyToken working in my example.

Then I think I don't understand what you mean. Can you give me a complete example that triggers this error? Because when I adjust your example the way I stated above I get this and this does not cause any error for `MyToken`:

```solidity
interface IERC20 {}
interface IOwnable {}
interface IERC20AltApprove {} // holds decreaseAllowance and increaseAllowance functions.
interface IERC20Metadata is IERC20 {} // optional metadata elements.
interface IERC20Burnable is IERC20Metadata {}
abstract contract Context {}
abstract contract Ownable is IOwnable, Context {}
contract ERC20 is IERC20Metadata, IERC20AltApprove, Context {}
contract ERC20Burnable is IERC20Burnable, Context, ERC20 {}

interface IMyToken is IOwnable, IERC20Burnable, IERC20AltApprove {}
contract MyToken is IMyToken, Ownable, ERC20Burnable {}
```

> Also, I am not sure if this is a separated feature or related. If you use interfaces in that way you came across cases where
an Interface holds a function that is defined in the interface and also already implemented in some base class.

Yes. There are reasons for this (see this topic on the forum if you're interested in details: [Thoughts on override(A, B) syntax](https://forum.soliditylang.org/t/thoughts-on-override-a-b-syntax/767)) but in the end we decided to relax the override restrictions for interfaces. This is tracked in #12665.

> Another drawback I came over my workaround is that while you can define functions and override them in the way shown you can not do the same thing with events.

You probably mean a situation like this?
```solidity
interface I {
    event E();
}

interface J {
    event E();
}

contract C is I, J {}
```
```
Error: Event with same name and parameter types defined twice.
 --> test.sol:6:5:
  |
6 |     event E();
  |     ^^^^^^^^^^
Note: Other declaration is here:
 --> test.sol:2:5:
  |
2 |     event E();
  |     ^^^^^^^^^^
```

Not sure if we'll do it but it's worth considering at least. Feel free to submit it as a separate issue with a feature request.

There's a technical limitation that prevents us from allowing situations where multiple events with the same signature are in scope: #11114. This can be solved but the solution requires bumping the Natspec version which would be a breaking change. But recently we have decided to do it a bit differently to make it non-breaking (#13086) so we might actually be in a position to allow this after all.

But then there's the question of whether it's a good idea. This needs some discussion. Providing a good, practical use case would help us make a decision.

**Comment by lukehutch** [comment]
I don't know if this is related, but I have the following issue:

```
interface A {
    function x() public returns (uint256);
}

contract B is A {
    uint256 public x;
}

contract C is A, B {
}
```

I get an error that contract C should be marked as `abstract` because it does not implement `function x()` from interface A (but it does implement this function, via the implicit function created from the inherited `uint256 public x` in contract `B`).

**Comment by cameel** [comment]
@lukehutch I'm not getting this error. I'm instead getting an error saying that `x()` should be `external` and making it `external` resolves it.

But yeah, this is completely unrelated to the issue here. If you're sure it's a problem in the compiler please report it as a separate issue and we'll investigate. But if you're not sure it's best to ask first, e.g. on the [#solidity](https://gitter.im/ethereum/solidity) channel first or on Ethereum Stack Exchange.

**Comment by migoldfinger** [comment]
> Then I think I don't understand what you mean. Can you give me a complete example that triggers this error? Because when I adjust your example the way I stated above I get this and this does not cause any error for `MyToken`:

@cameel Just give me some time until weekend and I will set up a repository that creates exactly the described problem.

> Yes. There are reasons for this (see this topic on the forum if you're interested in details: [Thoughts on override(A, B) syntax](https://forum.soliditylang.org/t/thoughts-on-override-a-b-syntax/767)) but in the end we decided to relax the override restrictions for interfaces. This is tracked in #12665.

Thanks for the clarification. In my case and opinion I would drop the override(A, B) syntax if every object except one is an interface. But this is not very important because it does not hurt to write the override(A, B) syntax. It just a nice to have.
```
interface IA { function f() external; }
interface IB { function f() external; }
contract A is IA, B { function f() override public{} } <- should be ok
contract A is IA, B { } <- should be ok also because f() it is already implemented in B
```
In every other case I would keep the override(A, B) syntax.

> You probably mean a situation like this?
> 
> ```solidity
> interface I {
>     event E();
> }
> 
> interface J {
>     event E();
> }
> 
> contract C is I, J {}
> ```

Exactly that's what pops up in my workaround attempt.

To sum it up I think that solidity handles interfaces to much like contracts.

**Comment by migoldfinger** [comment]
@cameel I tried for 2 hours now but was unable to recreate the "linearization of inheritance graph impossible" error.
Maybe you are right and I did have some mistake in the order somewhere or accidentally used the contract instead of the interface.
I will now start to remove the workaround and use my initial interface driven concept again. Maybe I ran into the error again but I do not believe it since I got a working prototype now.

**Comment by cameel** [comment]
Sure. If you manage to reproduce it, please report an issue. Best if you can actually get a short repro the size of the one you had above - that's all we need to investigate, it just needs to produce the error.

Also, I'll keep the issue open since I think you raised a very good point. Even if this is solvable by changing the order, ignoring interfaces in linearization still makes sense to me and would make the language easier to use. We'll consider it when we'll be overhauling inheritance. I'll just hide our discussion related to the debugging of the error to keep the focus on the feature here.

**Comment by migoldfinger** [comment]
@cameel I finally came across my linearization problem again.
I took the base contracts from https://github.com/Exoda/exoda-contracts/tree/interfaceextension the contracts are published and can be installed with `npm i @exoda/contracts`.
The contract in question Fermion.sol is here https://github.com/ExofiDEX/exofi-contracts/tree/dev.
I do have write access to both repositories.
I could not find any way to solve the linearization error on contracts/Fermion.sol:11:1: in any way.

As stated I could not find any order of listing the inherited interfaces/classes to avoid the error.
